test/e2e/sync_lb_fail_test.go

```diff
@@ -0,0 +1,125 @@
+		Selector: map[string]string{
+			"ingresscontroller.operator.openshift.io/deployment-ingresscontroller": icName.Name,
+		},
+		Ports: []corev1.ServicePort{
+			{
+				Name:       "http",
+				Port:       80,
+				Protocol:   corev1.ProtocolTCP,
+				TargetPort: intstr.IntOrString{IntVal: 80},
+			},
+			{
+				Name:       "https",
+				Port:       443,
+				Protocol:   corev1.ProtocolTCP,
+				TargetPort: intstr.IntOrString{IntVal: 443},
+			},
+		},
+	},
+}
+
+func TestSyncLBFailEvent(t *testing.T) {
+	if infraConfig.Status.Platform != configv1.AWSPlatformType {
+		t.Skip("test skipped on non-aws platforms")
+		return
+	}
+	if err := kclient.Create(context.TODO(), &badLBService); err != nil {
+		t.Fatalf("failed to create service: %v", err)
+	}
+	defer kclient.Delete(context.TODO(), &badLBService)
+	ic := newLoadBalancerController(icName, icName.Name+dnsConfig.Spec.BaseDomain)
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create ingresscontroller: %v", err)
+	}
+	defer kclient.Delete(context.TODO(), ic)
+	eventList := &corev1.EventList{}
+	err := wait.PollImmediate(1*time.Second, 5*time.Minute, func() (bool, error) {
+		if err := kclient.List(context.TODO(), eventList, client.InNamespace(badLBService.ObjectMeta.Namespace)); err != nil {
+			return false, nil
+		}
+		for _, event := range eventList.Items {
+			if event.Reason == "SyncLoadBalancerFailed" {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("Timed out waiting for SyncLoadBalancerFailed event")
+	}
+	expectedConditions := []operatorv1.OperatorCondition{
+		{Type: operatorv1.LoadBalancerReadyIngressConditionType, Status: operatorv1.ConditionFalse},
+		{Type: operatorv1.OperatorStatusTypeDegraded, Status: operatorv1.ConditionTrue},
+	}
+	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, icName, expectedConditions...); err != nil {
+		t.Fatalf("Timed out waiting for status conditions LoadBalancerReady=false and Degraded=true for ingresscontroller %s", icName.Name)
+	}
+	err = wait.PollImmediate(1*time.Second, 5*time.Minute, func() (bool, error) {
+		if err := kclient.List(context.TODO(), eventList, client.InNamespace(badLBService.ObjectMeta.Namespace)); err != nil {
+			return false, nil
+		}
+		for _, event := range eventList.Items {
+			if event.Reason == "SyncLoadBalancerFailed" {
+				kclient.Delete(context.TODO(), &event)
+			}
+		}
+		ic := &operatorv1.IngressController{}
+		if err := kclient.Get(context.TODO(), icName, ic); err != nil {
+			t.Logf("Failed to get ingresscontroller %s: %v", icName.Name, err)
+			return false, nil
+		}
+		expected := operatorConditionMap(expectedConditions...)
+		current := operatorConditionMap(ic.Status.Conditions...)
+		if conditionsMatchExpected(expected, current) {
+			return false, nil
+		} else {
+			return true, fmt.Errorf("ingresscontroller %s should remain degraded", icName.Name)
+		}
+	})
+	if err != nil && err != wait.ErrWaitTimeout {
```

Comment:
I don't think you can check for `ErrWaitTimeout` here.  You'll have to fix that error by making the time longer.

Reply:
When `wait.PollImmediate` returns an error, it always returns `wait.ErrWaitTimeout` if the timeout is reached. The idea of how I wrote this is that if `LoadBalancerReady` stays `False` and `Degraded` stays `True` for the entire timeout, it's considered a success, but if it changes from that status, the poll loop will exit early with a different error, indicating a failure.

Reply:
With these type of E2Es, I'm a little surprised we aren't concerned about ballooning the E2E test time, since this test guarantees 5 mins runtime on pass, but I supposed better safe than sorry.

Reply:
It's fine though, especially regarding my comment with parallelization.

Reply:
With parallelization, the few minutes here will almost definitely finish before our longest tests, but I reduced the time to 3 minutes to be a bit faster. It seems to still be long enough in the manual test runs I did.

Reply:
> 

I think you just need to return `true, nil` where appropriate.